Interfaces

AmmarAbudrhman opened the pull request.

Commit:
refactor(api): standardize endpoint routes to kebab-case

Commit:
feat(api): implement API versioning and integrate with Swagger UI

- Added Microsoft.AspNetCore.Mvc.Versioning and ApiExplorer
- Configured default API version and media type reader
- Enabled versioned Swagger documentation with dropdown for versions
- Kept JWT authentication support in Swagger

Commit:
The work you've done is excellent for creating clean and useful OpenAPI documentation. You've correctly used attributes to describe your API, which is a great practice.

Here's a message you can use to communicate what you've done, explaining the benefits and the specific attributes you've used.

OpenAPI Documentation Updates
I've updated the BlocksController.cs file to significantly improve our OpenAPI (Swagger) documentation. These changes will make our API clearer and easier for other developers to use.

Here's what I've done:

Endpoint-Specific Information: Each endpoint now includes a [SwaggerOperation] attribute with a clear Summary and Description. This helps developers quickly understand the purpose of each API call without having to read the code.

Explicit Responses: I've added [ProducesResponseType] attributes to define the exact HTTP status codes and data models that each endpoint can return. For example, Status200OK is used for successful requests, and Status404NotFound is used when a resource isn't found. This prevents guesswork and makes our API contract explicit.

Clear Parameter Descriptions: I used the [SwaggerParameter] attribute to provide a Description for each query string, route, and body parameter. This includes specifying whether a parameter is Required and clarifies what information the endpoint needs to work correctly.

Controller-Level Tags: I added [SwaggerTag("Blocks management endpoints")] to the controller. This will group our endpoints under a clear tag in the documentation, making navigation much simpler.

Commit:
Title:
Refactor Controllers and Services with Swagger, API Versioning, and Response Standardization

Description:
This PR includes comprehensive updates and refactoring of multiple controllers and services to standardize API design, improve documentation, and enforce consistent response handling across the application.

Changes included:

BlocksController

Added [Authorize] and [ValidateActionFilter] for security and validation.

Integrated Swagger documentation for endpoints with [SwaggerTag] and [SwaggerOperation].

Standardized HTTP response types using Response(await _BlockServices.XXXAsync(...)).

Added [MapToApiVersion("1.0")] for API versioning.

PersonController

Refactored to match BlocksController style with authorization, validation, and Swagger documentation.

Standardized all endpoints to return ApiResponse<T> consistently.

ProjectCatgoriesController & ProjectCatgoryService

Added structured responses and improved logging.

Standardized Add, Update, Delete, GetAll, and GetById methods.

ProjectsController & ProjectService

Updated endpoints to include assign/remove operations for teams and families.

Standardized response handling and parameter passing.

ProjectFamilyController & ProjectFamilieservice

Refactored for consistency with ApiResponse<T> pattern.

Added [FromBody] and [FromRoute] where appropriate.

TeamMembersController & TeamMemberService

Standardized CRUD endpoints.

Added [FromBody] for DTO inputs and consistent response handling.

TeamRoleController & TeamRoleService

Added [Authorize] and structured endpoints.

Standardized responses for add, update, delete, and retrieval operations.

TeamsController & TeamsService

Refactored to follow the same standardized pattern with Swagger documentation, authorization, and response handling.

Added endpoint for retrieving team projects.

Fixed HTTP verbs for proper REST conventions (POST for create, GET for retrieval, etc.).

Benefits:

Consistent API versioning and response structure.

Enhanced Swagger documentation for better API discoverability.

Improved security via [Authorize].

Reduced duplicated patterns and improved maintainability.

Testing:

All existing endpoints were tested manually for expected behavior.

Response messages are now consistent across all controllers.

CRUD operations for all entities (Block, Person, Project, Team, TeamRole) verified.

Commit:
Added new service with interface and registered in Dependency Injection

Created I<ServiceName> interface for abstraction.

Implemented business logic in <ServiceName>Service.

Registered the service in DI container (Program.cs) using AddScoped<I<ServiceName>, <ServiceName>Service>().

Grouped the service registration with related domain services for better organization.

Ensured controllers depend on the interface instead of the concrete implementation.

Dev7-Mubarak commented on SmartNeighborhoodAPI/Controllers/V1/BlocksController.cs:
```diff
@@ -0,0 +1,116 @@
+    [ApiController]
+    [ValidateActionFilter]
+    [ApiVersion("1.0")]
+    [Route("api/[controller]")]
```
Remove this form from all controllers because it already exists in `AppControllerBase`.

Dev7-Mubarak commented on SmartNeighborhoodAPI/Controllers/V1/BlocksController.cs:
```diff
@@ -0,0 +1,116 @@
+            _BlockServices = BlockServices;
+        }
+
+        [HttpGet("get-details")]
```
Good job, please move this to a separate class. Video link for reference is provided. https://youtu.be/oKn3Q0o_fhU?si=0k3pZsjxMd2_jJqR

Dev7-Mubarak commented:
Good job, but when I try to run the code it doesn't work. Please check everything is working as expected before opening a pull request.

Commit:
fix all the problem with the interfaces and remove unneccary things in the controller

Commit:
Delete the Ads And Group Controller and register all interfaces in the program

Commit:
Compelet the task succuflly check and give me your thought about it

Dev7-Mubarak committed:
Refactor API routing and update endpoint definitions

Updated routing constants in `Router.cs` to simplify endpoint paths for blocks, conflict cases, families, and other entities by removing action suffixes. Adjusted corresponding routes in controllers within the `SmartNeighborhoodAPI.Controllers.V1` namespace. Removed commented-out code in `AuthController` related to user registration and email confirmation. Modified `ChangeBlockManagerDto` to eliminate the `BlockId` property and updated method signatures in `IBlockServices` and `BlockServices` to accept an `id` parameter. These changes enhance consistency and usability of the API endpoints.

Dev7-Mubarak committed:
Refactor API routes for clarity and RESTful consistency

- Updated route prefixes in Router.cs for better structure and readability
- Renamed and corrected endpoint formats to follow RESTful conventions
- Modified ProjectController to use route parameters instead of query parameters for team and family assignments
- Refactored Teams routes and GetTeamProjects endpoint to use subresource style (e.g., /teams/{teamId}/projects)
- Enhanced API maintainability and usability by standardizing route naming and structure

Dev7-Mubarak committed:
Remove FamilyType entity and related configurations

This commit removes the `FamilyType` entity, its DbSet, and all related configurations from the codebase. The `Family` entity has been updated to make the `FamilyNotes` property nullable and to eliminate the foreign key reference to `FamilyType`. Migration files have been adjusted to drop the `FamilyTypes` table and update the database schema accordingly. All service classes and DTOs that referenced `FamilyType` have been modified to ensure consistency with the new data model. The `FamilyTypeService` class has also been removed as it is no longer needed.

Dev7-Mubarak committed:
Add AutoMapper config and update service registrations

This commit introduces AutoMapper configuration for the `Program` class and updates the dependency injection container. It retains JWT configuration and various service registrations while removing `IFamilyTypeService`, indicating a potential refactor or removal from the service layer.
